### PR TITLE
feat(LINGUIST-58): Added leaderboard

### PR DESCRIPTION
**Pull request by elif-krvan**

### CODE REVIEW

File: components/common/Card.tsx [hunk 1]
```diff
@@ -15,7 +15,7 @@ const styles = StyleSheet.create({
   card: {
     backgroundColor: 'white',
     borderRadius: 10,
-    shadowColor: Colors.primary[500],
```
Comment (selimcanglr) [inline components/common/Card.tsx:18]: I use this card in other places as well. Can you take this change back? You can always pass the style you wanted via the `style` prop.

File: components/common/Text.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+const LText = ({ children, size = 12, style, centered }: TextProps) => {
+  const textAlign = centered ? 'center' : 'left';
+  const fontsLoaded = useContext(FontLoadedContext);
+  let currentStyle = styles.titleTextCustom;
+
+  if (!fontsLoaded) {
+    currentStyle = styles.titleText;
+    return <Text style={[currentStyle, { fontSize: size, textAlign }, style]}>{children}</Text>;
```
Comment (selimcanglr) [inline components/common/Text.tsx:21]: You can remove the return in if statement, it is redundant. We return the same thing below anyways, setting `currentStyle` variable in the if statement seems enough.

File: components/common/Text.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+interface TextProps {
+  children: React.ReactNode;
+  size?: number;
+  style: StyleProp<TextStyle>;
+  centered?: boolean;
+}
+
+const LText = ({ children, size = 12, style, centered }: TextProps) => {
+  const textAlign = centered ? 'center' : 'left';
+  const fontsLoaded = useContext(FontLoadedContext);
```
Comment (selimcanglr) [inline components/common/Text.tsx:16]: A minor comment, could you try to keep the hook statements on top of the component in general. For example, we can carry this `useContext` to the top.

File: components/common/Title.tsx [hunk 4]
```diff
@@ -12,13 +15,28 @@ interface TitleProps {
 
 const Title = ({ children, size: fontSize = 'h1', centered }: TitleProps) => {
   const textAlign = centered ? 'center' : 'left';
-  return <Text style={[styles.titleText, { fontSize: TitleSizes[fontSize], textAlign }]}>{children}</Text>;
+  const fontsLoaded = useContext(FontLoadedContext);
+  let currentStyle = styles.titleTextCustom;
+
+  if (!fontsLoaded) {
+    currentStyle = styles.titleText;
+    return <Text style={[currentStyle, { fontSize: TitleSizes[fontSize], textAlign }]}>{children}</Text>;
```
Comment (selimcanglr) [inline components/common/Title.tsx:23]: Same comments for `LText` component applies here.

File: components/user/userApi.ts [hunk 5]
```diff
@@ -76,6 +85,22 @@ export const userApi = createApi({
       }),
       invalidatesTags: ['FriendRequest'],
     }),
+    getGlobalLeaderboard: builder.query<RLeaderboard, QLeaderboard>({
+      query: (paginationParams) => ({
+        url: '/leaderboard/global/xp',
+        method: 'GET',
+        params: paginationParams,
+      }),
+      providesTags: ['Leaderboard'],
```
Comment (selimcanglr) [inline components/user/userApi.ts:94]: Do we invalıdate these in any other place? If not, we don't need to provide these tags.

File: components/user/userApi.ts [hunk 6]
```diff
@@ -76,6 +85,22 @@ export const userApi = createApi({
...
+    }),
+    getFriendLeaderboard: builder.query<RLeaderboard, QLeaderboard>({
+      query: (paginationParams) => ({
+        url: '/leaderboard/friends/xp',
+        method: 'GET',
+        params: paginationParams,
+      }),
+      providesTags: ['FriendLeaderboard'],
```
Comment (selimcanglr) [inline components/user/userApi.ts:102]: Same as above.

File: screens/leaderboard/FriendsLeaderboardScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+  useEffect(() => {
+    trigger({ size: DEFAULT_PAGE_SIZE });
+    // setCurrentPage(leaderboard!.currentPage);
+  }, []);
+
+  useEffect(() => {
+    if (leaderboard) {
+      setCurrentPage(leaderboard.currentPage);
+      setTotalPageNum(leaderboard.totalPages);
```
Comment (selimcanglr) [inline screens/leaderboard/FriendsLeaderboardScreen.tsx:44]: Handling these stuff in a `useEffect` is not suggested. Also, we don't need to keep these as a separate state anyways. `leaderboard` variable already contains these information we can have these as variables.
Just add the following lines after `useLazyGetFriendLeaderboardQuery`:
```js
const currentPage = leaderboard?.currentPage || DEFAULT_PAGE;
const totalPageNum = leaderboard?.totalPages || 0;
```


And then remove the states for `currentPage` and `totalPageNum`. Also please remove the commented state changes.

File: screens/leaderboard/GlobalLeaderboardScreen.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+const DEFAULT_PAGE_SIZE = 10;
+
+const GlobalLeaderboardScreen = () => {
+  const [currentPage, setCurrentPage] = useState(DEFAULT_PAGE);
+  const [totalPageNum, setTotalPageNum] = useState(0);
+  const [isRefreshing, setIsRefreshing] = useState(false);
+  const [trigger, { data: leaderboard, isFetching, isError }] = useLazyGetGlobalLeaderboardQuery();
+
+  const goToPreviousPage = () => {
+    if (currentPage > 0) {
```
Comment (selimcanglr) [inline screens/leaderboard/GlobalLeaderboardScreen.tsx:17]: Comments for the `FriendLeaderboardScreen` apply here as well. Please apply the same changes.

**Review by selimcanglr — CHANGES_REQUESTED** [review]

File: components/common/Text.tsx (hunk 2 above)
Comment (elif-krvan) [reply components/common/Text.tsx:21]: bunu circle oluyo diye warning verdiği için eklemiştim bi şey değişmedi ama. En son böyle unutmuşum

File: components/common/Text.tsx [hunk 9]
```diff
@@ -1,41 +1,40 @@
...
-  if (!fontsLoaded) {
-    currentStyle = styles.titleText;
-    return <Text style={[currentStyle, { fontSize: size, textAlign }, style]}>{children}</Text>;
+  if (style && (style as TextStyle).fontWeight === 'bold') {
+    const { fontWeight, ...rest } = style as TextStyle;
+    currentStyle = {
+      ...styles.titleTextCustomBolder,
+      ...rest,
+    };
+    console.log(rest);
```
Comment (selimcanglr) [inline components/common/Text.tsx]: There is a console.log here

File: package.json [hunk 10]
```diff
@@ -56,12 +60,10 @@
     "react-native-safe-area-context": "4.6.3",
     "react-native-screens": "~3.22.0",
     "react-native-shimmer-placeholder": "^2.0.9",
-    "react-native-svg": "13.9.0",
+    "react-native-svg": "^15.1.0",
```
Comment (selimcanglr) [inline package.json:63]: Can you run expo install --fix for this one

**Review by selimcanglr — APPROVED** [review]

**Pull request merged**